File: analysis/revision/scripts/statistical_tests.py
# Statistical tests supporting the revision:
#  1. Fisher exact tests on the incipient site contrast (by threshold; generalists-only)
#  2. Calibration-anchor uncertainty: CI on the mean calibration line vs single-tree PI
# Run from repo root.
import pandas as pd, numpy as np
from scipy import stats

df = pd.read_csv('data/Tree_ID_info.csv'); df['site'] = np.where(df['plot'] == 'BGS', 'BGS', 'EMS')
sm = {'rm': 'A.rubrum', 'bg': 'N.sylvatica', 'ro': 'Q.rubra', 'hem': 'T.canadensis'}
df['sp'] = df['species'].map(sm)
ert = pd.read_csv('data/ERT_application_results.csv')
m = df.merge(ert, on='tree').reset_index(drop=True)
mets = ['mean', 'median', 'sd', 'cv', 'gini', 'entropy', 'cma', 'radialgradient']
Z = m[mets].copy()
for c in mets:
    Z[c] = m.groupby('sp')[c].transform(lambda x: (x - x.mean()) / x.std(ddof=1))
Zc = Z.values - Z.values.mean(0)
U, S, Vt = np.linalg.svd(Zc, full_matrices=False)
pc = (U * S)[:, 0]
if np.corrcoef(pc, m['mean'])[0, 1] > 0:
    pc = -pc
m['pc1'] = pc; m['struct'] = m['percent_damaged'] > 1
sd = pc.std(ddof=1); sound = ~m['struct']

print("1. Fisher exact: incipient site contrast by ERT threshold")
for k, lab in [(-0.5, 'mean-0.5SD'), (0, 'mean (published)'), (0.5, 'mean+0.5SD'), (1, 'mean+1SD')]:
    t = pc.mean() + k * sd; inc = sound & (pc > t)
    a = int(inc[m.site == 'BGS'].sum()); b = int((sound & (m.site == 'BGS') & ~inc).sum())
    c = int(inc[m.site == 'EMS'].sum()); d = int((sound & (m.site == 'EMS') & ~inc).sum())
    OR, p = stats.fisher_exact([[a, b], [c, d]])
    print(f"  {lab:18s}: BGS {a}/{a+b} vs EMS {c}/{c+d}   OR={OR:.2f}  p={p:.3f}")

inc0 = sound & (pc > pc.mean())
a = int(inc0[m.site == 'BGS'].sum()); n1 = int((m.site == 'BGS').sum())
c = int(inc0[m.site == 'EMS'].sum()); n2 = int((m.site == 'EMS').sum())
OR, p = stats.fisher_exact([[a, n1 - a], [c, n2 - c]])
print(f"  As % of ALL trees (31% vs 11%): {a}/{n1} vs {c}/{n2}  OR={OR:.2f}  p={p:.3f}")

g = m[m.sp.isin(['A.rubrum', 'T.canadensis'])]
incg = (~g['struct']) & (g['pc1'] > pc.mean())
a = int(incg[g.site == 'BGS'].sum()); n1 = int((g.site == 'BGS').sum())
c = int(incg[g.site == 'EMS'].sum()); n2 = int((g.site == 'EMS').sum())
OR, p = stats.fisher_exact([[a, n1 - a], [c, n2 - c]])
print(f"  GENERALISTS ONLY: BGS {a}/{n1} ({a/n1*100:.0f}%) vs EMS {c}/{n2} ({c/n2*100:.0f}%)  OR={OR:.2f}  p={p:.3f}")

print("\n2. Absolute anchor uncertainty (hemlock calibration, n=12)")
v = pd.read_csv('data/hemlock/validation_summary.csv'); v['res'] = 1000 / v['Conductance']
sl = stats.linregress(v['res'], v['moisture'])
n = len(v); xbar = v['res'].mean(); Sxx = ((v['res'] - xbar) ** 2).sum()
rstd = np.sqrt((((v['moisture'] - (sl.intercept + sl.slope * v['res']))) ** 2).sum() / (n - 2))
tcrit = stats.t.ppf(0.975, n - 2)
for M in [90, 100, 110]:
    x = (M - sl.intercept) / sl.slope
    ci = rstd * np.sqrt(1 / n + (x - xbar) ** 2 / Sxx) / abs(sl.slope) * tcrit
    pi = rstd * np.sqrt(1 + 1 / n + (x - xbar) ** 2 / Sxx) / abs(sl.slope) * tcrit
    print(f"  moisture>{M}% -> res<{x:.0f} Ohm-m; 95% CI (stand-level anchor): {x-ci:.0f}..{x+ci:.0f}; "
          f"95% PI (single tree): {x-pi:.0f}..{x+pi:.0f}")

File: analysis/revision/scripts/statistical_tests.R
# Statistical tests supporting the revision (R port of statistical_tests.py):
#  1. Fisher exact tests on the incipient site contrast (by threshold; generalists-only)
#  2. Calibration-anchor uncertainty: CI on the mean calibration line vs single-tree PI
# Run from repo root: Rscript analysis/revision/scripts/statistical_tests.R
source("analysis/revision/scripts/revision_common.R")

m <- load_merged()
bp <- build_pc1(m, "species")
pc <- bp$pc1
m$pc1 <- pc
m$struct <- m$percent_damaged > 1
sdv <- sd(pc)          # ddof=1, as in the Python script
sound <- !m$struct

# Sample odds ratio (a*d)/(b*c), matching scipy.stats.fisher_exact's OR;
# p-value from R's fisher.test (identical two-sided p).
fisher_line <- function(a, b, c, d) {
  or <- (a * d) / (b * c)
  p <- fisher.test(matrix(c(a, b, c, d), nrow = 2, byrow = TRUE))$p.value
  list(or = if (is.finite(or)) sprintf("%.2f", or) else "inf", p = p)
}

cat("1. Fisher exact: incipient site contrast by ERT threshold\n")
cases <- list(list(-0.5, "mean-0.5SD"), list(0, "mean (published)"),
              list(0.5, "mean+0.5SD"), list(1, "mean+1SD"))
for (cs in cases) {
  t0 <- mean(pc) + cs[[1]] * sdv
  inc <- sound & (pc > t0)
  a <- sum(inc & m$site == "BGS"); b <- sum(sound & m$site == "BGS" & !inc)
  c <- sum(inc & m$site == "EMS"); d <- sum(sound & m$site == "EMS" & !inc)
  f <- fisher_line(a, b, c, d)
  cat(sprintf("  %-18s: BGS %d/%d vs EMS %d/%d   OR=%s  p=%.3f\n",
              cs[[2]], a, a + b, c, c + d, f$or, f$p))
}

inc0 <- sound & (pc > mean(pc))
a <- sum(inc0 & m$site == "BGS"); n1 <- sum(m$site == "BGS")
c <- sum(inc0 & m$site == "EMS"); n2 <- sum(m$site == "EMS")
f <- fisher_line(a, n1 - a, c, n2 - c)
cat(sprintf("  As %% of ALL trees (31%% vs 11%%): %d/%d vs %d/%d  OR=%s  p=%.3f\n",
            a, n1, c, n2, f$or, f$p))

g <- m[m$sp %in% c("A.rubrum", "T.canadensis"), ]
incg <- (!g$struct) & (g$pc1 > mean(pc))
a <- sum(incg & g$site == "BGS"); n1 <- sum(g$site == "BGS")
c <- sum(incg & g$site == "EMS"); n2 <- sum(g$site == "EMS")
f <- fisher_line(a, n1 - a, c, n2 - c)
cat(sprintf("  GENERALISTS ONLY: BGS %d/%d (%.0f%%) vs EMS %d/%d (%.0f%%)  OR=%s  p=%.3f\n",
            a, n1, a / n1 * 100, c, n2, c / n2 * 100, f$or, f$p))

cat("\n2. Absolute anchor uncertainty (hemlock calibration, n=12)\n")
v <- load_validation()
res <- v$mean_res
fit <- lm(moisture ~ res, data = data.frame(res = res, moisture = v$moisture))
intercept <- coef(fit)[1]; slope <- coef(fit)[2]
n <- nrow(v); xbar <- mean(res); Sxx <- sum((res - xbar)^2)
rstd <- sqrt(sum(residuals(fit)^2) / (n - 2))
tcrit <- qt(0.975, n - 2)
for (M in c(90, 100, 110)) {
  x <- (M - intercept) / slope
  ci <- rstd * sqrt(1 / n + (x - xbar)^2 / Sxx) / abs(slope) * tcrit
  pi <- rstd * sqrt(1 + 1 / n + (x - xbar)^2 / Sxx) / abs(slope) * tcrit
  cat(sprintf("  moisture>%d%% -> res<%.0f Ohm-m; 95%% CI (stand-level anchor): %.0f..%.0f; 95%% PI (single tree): %.0f..%.0f\n",
              M, x, x - ci, x + ci, x - pi, x + pi))
}
